File: app/src/main/java/com/example/myapplication/guardian/userchat/ui/UserChatActivity.java
package com.example.myapplication.guardian.userchat.ui;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.media.MediaRecorder;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myapplication.R;
import com.example.myapplication.common.model.ChatMessage;

import org.json.JSONArray;
import org.json.JSONObject;
import org.vosk.Model;
import org.vosk.Recognizer;
import org.vosk.android.RecognitionListener;
import org.vosk.android.SpeechService;
import org.vosk.android.StorageService;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.*;

public class UserChatActivity extends AppCompatActivity implements ChatAdapter.OnSuggestionClickListener, RecognitionListener {

    // 서버 URL (스크린샷 기준)
    private final String SERVER_URL = "http://10.100.0.51:8080/nungil-server/api/v1/nungil/analyze";
    private final OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build();

    private List<ChatMessage> chatList = new ArrayList<>();
    private ChatAdapter adapter;
    private RecyclerView rvChat;
    private ProgressBar loadingBar;

    private SpeechService speechService;
    private Model model;
    private MediaRecorder recorder;
    private File voiceFile;
    private boolean isRecording = false;
    private int voiceMsgIndex = -1;

    // [핵심] 새로운 사진이 찍히기 전까지 유지될 비트맵 변수
    private Bitmap lastCapturedBitmap = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user_chat);

        rvChat = findViewById(R.id.rvChat);
        loadingBar = findViewById(R.id.loadingBar);
        adapter = new ChatAdapter(chatList, this);

        // 리사이클러뷰 설정 (Padding과 clipToPadding은 XML에서 설정 권장)
        rvChat.setLayoutManager(new LinearLayoutManager(this));
        rvChat.setAdapter(adapter);

        checkPermissions();

        // 카메라 버튼
        findViewById(R.id.btnCamera).setOnClickListener(v ->
                startActivityForResult(new Intent(MediaStore.ACTION_IMAGE_CAPTURE), 1));

        // 마이크 버튼
        findViewById(R.id.btnNext).setOnClickListener(v -> {
            if (!isRecording && loadingBar.getVisibility() != View.VISIBLE) {
                Toast.makeText(this, "어떤 것이 궁금하신가요? 말씀해주세요!", Toast.LENGTH_SHORT).show();
                stopVosk();
                startVoiceRecording();
            }
        });

        addMsg("반가워요! 사진을 찍거나 '똘똘아'라고 불러보세요.", ChatMessage.TYPE_OTHER, false, null, null);
    }

    /**
     * 서버 전송 로직: 새로운 사진이 없으면 lastCapturedBitmap을 재사용하여 전송
     */
    private void uploadToServer(@Nullable Bitmap newBitmap, @Nullable File voice, @Nullable String text) {
        stopVosk(); // 서버 요청 중 호출어 인식 방지
        runOnUiThread(() -> loadingBar.setVisibility(View.VISIBLE));

        MultipartBody.Builder builder = new MultipartBody.Builder().setType(MultipartBody.FORM);

        // 1. 공통 ID 정보
        builder.addFormDataPart("info", "{\"id\": \"ttoli_user_01\"}");

        // 2. 텍스트 정보 (말한 내용 혹은 추천 질문)
        if (text != null) builder.addFormDataPart("text", text);

        // 3. 음성 파일 정보
        if (voice != null) {
            builder.addFormDataPart("voice", "voice_input.m4a",
                    RequestBody.create(voice, MediaType.parse("audio/m4a")));
        }

        // 4. [이미지 컨텍스트 로직]
        if (newBitmap != null) {
            // 새로 사진을 찍은 경우: 변수를 갱신하고 새 사진을 보냄
            lastCapturedBitmap = newBitmap;
            builder.addFormDataPart("image", "current_shot.jpg",
                    RequestBody.create(bitmapToFile(newBitmap), MediaType.parse("image/jpeg")));
        } else if (lastCapturedBitmap != null) {
            // 새로 찍은 사진은 없지만, 기존에 찍어둔 사진이 있는 경우: 기존 사진을 다시 보냄
            builder.addFormDataPart("image", "current_shot.jpg",
                    RequestBody.create(bitmapToFile(lastCapturedBitmap), MediaType.parse("image/jpeg")));
        }

        Request request = new Request.Builder().url(SERVER_URL).post(builder.build()).build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NonNull Call call, @NonNull IOException e) {
                runOnUiThread(() -> {
                    loadingBar.setVisibility(View.GONE);
                    updateVoiceStatus("❌ 서버 연결 실패");
                    resetToVoskMode();
                });
            }

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
                String body = response.body().string();
                new Handler(Looper.getMainLooper()).post(() -> {
                    loadingBar.setVisibility(View.GONE);
                    if (response.isSuccessful()) {
                        parseServerResponse(body);
                    } else {
                        updateVoiceStatus("❌ 서버 응답 오류");
                    }
                    resetToVoskMode();
                });
            }
        });
    }

    private void parseServerResponse(String json) {
        try {
            JSONObject root = new JSONObject(json);
            if ("SUCCESS".equals(root.optString("status"))) {
                JSONObject result = root.getJSONObject("result");

                // transcribedText 업데이트
                String transcribed = result.optString("transcribedText", "");
                if (!transcribed.isEmpty()) {
                    updateVoiceStatus(transcribed);
                    voiceMsgIndex = -1;
                }

                // 답변 및 추천 질문 추가
                String answer = result.optString("answer", "");
                List<String> suggestions = new ArrayList<>();
                JSONArray suggestArray = result.optJSONArray("suggestedQuestions");
                if (suggestArray != null) {
                    for (int i = 0; i < suggestArray.length(); i++) {
                        suggestions.add(suggestArray.getString(i));
                    }
                }
                addMsg(answer, ChatMessage.TYPE_OTHER, false, null, suggestions);
            }
        } catch (Exception e) {
            updateVoiceStatus("❌ 데이터 처리 오류");
        }
    }

    // --- 호출어 감지 (Vosk) 파트 ---

    @Override
    public void onResult(String hypothesis) {
        try {
            JSONObject json = new JSONObject(hypothesis);
            String text = json.optString("text", "").trim();

            if (text.isEmpty() || isRecording || loadingBar.getVisibility() == View.VISIBLE) return;

            // [잡음 방지] 딱 "똘똘" 혹은 "똘똘아"만 정확히 인식되었을 때(글자수 제한 및 정규식)만 실행
            if (text.matches("^(똘똘|똘똘아)$")) {
                Log.d("Vosk", "호출 감지: " + text);
                runOnUiThread(() -> {
                    Toast.makeText(this, "네, 말씀하세요!", Toast.LENGTH_SHORT).show();
                    stopVosk();
                    startVoiceRecording();
                });
            }
        } catch (Exception e) { e.printStackTrace(); }
    }

    @Override public void onPartialResult(String s) { /* 오작동 방지를 위해 비워둠 */ }

    // --- 녹음 로직 ---

    private void startVoiceRecording() {
        isRecording = true;
        runOnUiThread(() -> {
            chatList.add(new ChatMessage("🎤 듣고 있습니다...", ChatMessage.TYPE_MINE, false, null, null));
            voiceMsgIndex = chatList.size() - 1;
            adapter.notifyItemInserted(voiceMsgIndex);
            rvChat.smoothScrollToPosition(voiceMsgIndex);
        });

        try {
            voiceFile = new File(getExternalFilesDir(null), "voice_input.m4a");
            recorder = new MediaRecorder();
            recorder.setAudioSource(MediaRecorder.AudioSource.MIC);
            recorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
            recorder.setAudioEncoder(MediaRecorder.AudioEncoder.AAC);
            recorder.setOutputFile(voiceFile.getAbsolutePath());
            recorder.prepare();
            recorder.start();
            new Handler(Looper.getMainLooper()).postDelayed(this::stopAndUploadVoice, 5000);
        } catch (Exception e) { resetToVoskMode(); }
    }

    private void stopAndUploadVoice() {
        if (recorder != null && isRecording) {
            try { recorder.stop(); recorder.release(); } catch (Exception e) {}
            recorder = null; isRecording = false;
            updateVoiceStatus("🔍 분석 중...");
            uploadToServer(null, voiceFile, null); // 녹음본 전송 (기존 이미지가 있다면 함께 감)
        }
    }

    private void updateVoiceStatus(String newText) {
        runOnUiThread(() -> {
            if (voiceMsgIndex != -1 && voiceMsgIndex < chatList.size()) {
                chatList.get(voiceMsgIndex).setContent(newText);
                adapter.notifyItemChanged(voiceMsgIndex);
            }
        });
    }

    // --- 엔진 및 기타 설정 ---

    private void startVosk() {
        try {
            if (speechService != null || isRecording) return;
            // 인식 범위를 똘똘, 똘똘아로 좁혀서 잡음 간섭 최소화
            Recognizer rec = new Recognizer(model, 16000.0f, "[\"똘똘\", \"똘똘아\"]");
            speechService = new SpeechService(rec, 16000.0f);
            speechService.startListening(this);
        } catch (Exception e) { e.printStackTrace(); }
    }

    private void stopVosk() { if (speechService != null) { speechService.stop(); speechService.shutdown(); speechService = null; } }

    private void resetToVoskMode() { isRecording = false; if (recorder != null) { try { recorder.release(); } catch(Exception e){} } recorder = null; startVosk(); }

    private void checkPermissions() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.RECORD_AUDIO, Manifest.permission.CAMERA}, 100);
        } else { initVosk(); }
    }

    private void initVosk() {
        StorageService.unpack(this, "model-ko", "model", (m) -> { this.model = m; startVosk(); }, (e) -> Log.e("Vosk", "Load Fail"));
    }

    private void addMsg(String c, int t, boolean i, Bitmap b, List<String> s) {
        runOnUiThread(() -> {
            chatList.add(new ChatMessage(c, t, i, b, s));
            adapter.notifyItemInserted(chatList.size()-1);
            rvChat.smoothScrollToPosition(chatList.size()-1);
        });
    }

    @Override public void onSuggestionClick(String t) {
        if (loadingBar.getVisibility() == View.VISIBLE) return;
        addMsg(t, ChatMessage.TYPE_MINE, false, null, null);
        uploadToServer(null, null, t); // 추천 질문 클릭 시에도 기존 사진이 있으면 함께 감
    }

    @Override protected void onActivityResult(int req, int res, @Nullable Intent d) {
        super.onActivityResult(req, res, d);
        if (res == RESULT_OK && req == 1 && d != null) {
            Bitmap b = (Bitmap) d.getExtras().get("data");
            addMsg(null, ChatMessage.TYPE_MINE, true, b, null);
            uploadToServer(b, null, null); // 새 사진 촬영 및 전송 (lastCapturedBitmap 갱신)
        }
    }

    private File bitmapToFile(Bitmap b) {
        File f = new File(getExternalFilesDir(null), "current_shot.jpg");
        try (FileOutputStream o = new FileOutputStream(f)) { b.compress(Bitmap.CompressFormat.JPEG, 90, o); } catch (Exception e) {}
        return f;
    }

    @Override public void onError(Exception e) { resetToVoskMode(); }
    @Override public void onTimeout() { resetToVoskMode(); }
    @Override public void onFinalResult(String s) { }
    @Override protected void onDestroy() { super.onDestroy(); stopVosk(); if (recorder != null) recorder.release(); }
}

File: app/src/main/java/com/example/myapplication/user/chat/UserChatActivity.kt
package com.example.myapplication.user.chat
import android.Manifest
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.MediaStore
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.user.UserStartActivity
import com.example.myapplication.R
import com.example.myapplication.common.model.UserChatLog
import com.example.myapplication.common.model.UserChatMessage
import com.example.myapplication.config.AppConfig
import com.example.myapplication.core.manager.TTSManager
import com.example.myapplication.core.manager.VoiceRecorderManager
import com.example.myapplication.core.manager.VoskWakeWordManager
import com.example.myapplication.user.schedule.data.ScheduleRepository
import com.example.myapplication.user.schedule.service.ScheduleAlarmReceiver
import com.example.myapplication.user.schedule.service.ScheduleManager
import com.google.gson.Gson
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit

class UserChatActivity : AppCompatActivity(), ChatAdapter.OnSuggestionClickListener {

    private val SERVER_URL = AppConfig.BASE_URL + "api/v1/question/analyze"
    private lateinit var USER_ID: String
    private var USER_IDX: Int = 1

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private val chatList: MutableList<UserChatMessage> = mutableListOf()
    private val conversationHistory: MutableList<UserChatLog> = mutableListOf()
    private val gson = Gson()

    private lateinit var adapter: ChatAdapter
    private lateinit var rvChat: RecyclerView
    private lateinit var loadingBar: ProgressBar
    private lateinit var btnMic: View
    private lateinit var ivBear: ImageView

    private enum class BearMood { BASIC, WORRY, PRAISE }

    private lateinit var voiceRecorder: VoiceRecorderManager
    private lateinit var voskManager: VoskWakeWordManager
    private lateinit var ttsManager: TTSManager
    private lateinit var scheduleManager: ScheduleManager

    private var isRecording = false
    private var voiceMsgIndex = -1
    private var shouldLaunchCamera = false
    private val mainHandler = Handler(Looper.getMainLooper())

    private var userSpecialNote = ""
    private var whitelistTaskNames = listOf<String>()

    private var scheduleTitle = ""
    private var scheduleSteps = listOf<String>()
    private var scheduleNote = ""
    private var currentScheduleId = -1
    private var currentStepIndex = -1
    private val isScheduleMode get() = currentStepIndex >= 0

    private val scheduleReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val scheduleId = intent.getStringExtra("schedule_id") ?: return
            val title = intent.getStringExtra("schedule_title") ?: "일정"
            showScheduleDialog(scheduleId, title)
        }
    }

    private val takePictureLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == RESULT_OK) {
            val bitmap = result.data?.extras?.get("data") as? Bitmap
            bitmap?.let {
                addMsg(null, UserChatMessage.TYPE_MINE, true, it, null)
                uploadToServer(it, null, null)
            }
        }
        resetToIdleState()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_user_chat)

        val prefs = getSharedPreferences("nungil_prefs", Context.MODE_PRIVATE)
        val userId = prefs.getString("user_id", null)
        if (userId == null) {
            startActivity(Intent(this, UserStartActivity::class.java))
            finish()
            return
        }
        USER_ID = userId
        USER_IDX = prefs.getInt("user_idx", 1)

        scheduleManager = ScheduleManager(this)
        ScheduleManager.createNotificationChannel(this)

        setupUI()
        setupManagers()
        checkPermissions()

        val welcomeMsg = "반가워요! '똘똘아'라고 불러보세요."
        addMsg(welcomeMsg, UserChatMessage.TYPE_OTHER, false, null, mutableListOf("오늘 날씨 어때?", "넌 누구니?"))
        conversationHistory.add(UserChatLog("AI", welcomeMsg))
        mainHandler.postDelayed({ ttsManager.speak(welcomeMsg) }, 1000)

        loadUserDataFromDB()
        handleScheduleIntent(intent)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        handleScheduleIntent(intent)
    }

    private fun loadUserDataFromDB() {
        Thread {
            val userInfo = ScheduleRepository.fetchUserInfo(USER_ID, USER_IDX)
            runOnUiThread {
                userSpecialNote = userInfo.specialNote
                whitelistTaskNames = userInfo.whitelistTaskNames
            }
        }.start()
        scheduleManager.syncSchedulesFromDB(userId = USER_ID, userIdx = USER_IDX)
    }

    private fun setupUI() {
        rvChat = findViewById(R.id.rvChat)
        loadingBar = findViewById(R.id.loadingBar)
        btnMic = findViewById(R.id.btnNext)
        ivBear = findViewById(R.id.ivBear)

        adapter = ChatAdapter(chatList, this)
        rvChat.layoutManager = LinearLayoutManager(this)
        rvChat.adapter = adapter

        btnMic.setOnClickListener {
            if (loadingBar.visibility == View.VISIBLE) return@setOnClickListener

            when {
                ttsManager.isSpeaking() -> {
                    ttsManager.stop()
                    handleTtsEndFlow()
                }
                isRecording -> {
                    voiceRecorder.cancel()
                    resetToIdleState()
                }
                else -> {
                    startRecordingFlow(manual = true)
                }
            }
        }
    }

    private fun handleTtsEndFlow() {
        runOnUiThread {
            updateMicButtonUI(forceMic = true)
            if (shouldLaunchCamera) {
                shouldLaunchCamera = false
                mainHandler.postDelayed({ openBackCamera() }, 300)
            } else {
                if (!isRecording && loadingBar.visibility != View.VISIBLE) voskManager.startListening()
            }
        }
    }

    private fun updateMicButtonUI(forceMic: Boolean = false) {
        runOnUiThread {
            val mic = btnMic as? ImageView ?: return@runOnUiThread
            val (iconRes, bgColor) = when {
                ttsManager.isSpeaking() && !forceMic -> android.R.drawable.ic_media_next to Color.TRANSPARENT
                isRecording && !forceMic -> android.R.drawable.ic_menu_close_clear_cancel to Color.parseColor("#FF5252")
                else -> android.R.drawable.ic_btn_speak_now to Color.TRANSPARENT
            }
            mic.setImageResource(iconRes)
            btnMic.setBackgroundColor(bgColor)
        }
    }

    private fun setupManagers() {
        ttsManager = TTSManager(this)
        ttsManager.onStatusListener = { isSpeaking ->
            if (!isSpeaking) handleTtsEndFlow() else updateMicButtonUI(forceMic = false)
        }

        voiceRecorder = VoiceRecorderManager(this, object : VoiceRecorderManager.RecorderListener {
            override fun onRecordingStart() {
                isRecording = true
                runOnUiThread {
                    updateMicButtonUI()
                    addMsg("🎤 듣고 있습니다...", UserChatMessage.TYPE_MINE, false, null, null)
                    voiceMsgIndex = chatList.size - 1
                }
            }
            override fun onRecordingSuccess(file: File) {
                isRecording = false
                updateMicButtonUI()
                updateVoiceStatus("🔍 분석 중...")
                uploadToServer(null, file, null)
            }
            override fun onRecordingCancel() { resetToIdleState() }
            override fun onError(message: String) { resetToIdleState() }
        })

        voskManager = VoskWakeWordManager(this, "[\"똘똘\", \"똘똘아\"]", object : VoskWakeWordManager.WakeWordListener {
            override fun onKeywordDetected() {
                if (loadingBar.visibility != View.VISIBLE && !isRecording && !ttsManager.isSpeaking())
                    startRecordingFlow(manual = false)
            }
            override fun onModelLoaded() { voskManager.startListening() }
            override fun onModelLoadFail() { Log.e("VOSK", "Load Fail") }
        })
    }

    private fun uploadToServer(newBitmap: Bitmap?, voice: File?, text: String?) {
        runOnUiThread {
            loadingBar.visibility = View.VISIBLE
            setBearMood(BearMood.WORRY)
            voskManager.stopListening()
        }

        val finalText = if (text != null && !isScheduleMode) "${buildUserContext()}\n\n사용자 질문: $text" else text
        if (finalText != null) conversationHistory.add(UserChatLog("user", finalText))

        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
        val textMt = "text/plain; charset=utf-8".toMediaTypeOrNull()

        builder.addFormDataPart("userId", null, USER_ID.toRequestBody(textMt))
        builder.addFormDataPart("historyJson", null, gson.toJson(conversationHistory).toRequestBody(textMt))
        finalText?.let { builder.addFormDataPart("textPrompt", null, it.toRequestBody(textMt)) }

        newBitmap?.let { bitmap ->
            val stream = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, 80, stream)
            builder.addFormDataPart("imageFile", "capture.jpg", stream.toByteArray().toRequestBody("image/jpeg".toMediaTypeOrNull()))
        }
        voice?.let { if (it.exists()) builder.addFormDataPart("voiceFile", it.name, it.asRequestBody("audio/m4a".toMediaTypeOrNull())) }

        val request = Request.Builder().url(SERVER_URL).post(builder.build()).build()
        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                Log.e("ServerCheck", "네트워크 오류 발생: ${e.message}")
                runOnUiThread { loadingBar.visibility = View.GONE; setBearMood(BearMood.BASIC); resetToIdleState() }
            }

            override fun onResponse(call: Call, response: Response) {
                val body = response.body?.string() ?: ""

                // [로그 추가] 서버 응답 원본 확인
                Log.d("ServerCheck", "── 서버 응답 수신 ──")
                Log.d("ServerCheck", "코드: ${response.code}")
                Log.d("ServerCheck", "본문: $body")
                Log.d("ServerCheck", "──────────────────")

                runOnUiThread {
                    loadingBar.visibility = View.GONE
                    setBearMood(BearMood.BASIC)
                    if (response.isSuccessful) {
                        parseServerResponse(body)
                    } else {
                        Log.e("ServerCheck", "서버 응답 실패: ${response.code}")
                        resetToIdleState()
                    }
                }
            }
        })
    }

    private fun parseServerResponse(json: String) {
        try {
            val root = JSONObject(json)
            if (root.optString("status") == "SUCCESS") {
                val result = root.getJSONObject("result")
                shouldLaunchCamera = result.optBoolean("photoRequest", false)

                val transcribed = result.optString("transcribedText", "")
                if (transcribed.isNotEmpty()) {
                    updateVoiceStatus(transcribed)
                    conversationHistory.add(UserChatLog("user", transcribed))
                }

                val answer = result.optString("answer", "")
                conversationHistory.add(UserChatLog("AI", answer))

                val suggestArray = result.optJSONArray("suggestedQuestions")
                val suggests = mutableListOf<String?>()
                if (suggestArray != null) {
                    for (i in 0 until suggestArray.length()) suggests.add(suggestArray.getString(i))
                }

                if (isScheduleMode && scheduleSteps.isNotEmpty()) {
                    val nextLabel = if (currentStepIndex + 1 < scheduleSteps.size) "다음 단계로" else "완료"
                    suggests.add(0, nextLabel)
                }

                val userContent = conversationHistory.lastOrNull { it.role == "user" }?.message ?: ""
                Thread { ScheduleRepository.logQuestionSuccess(ScheduleRepository.logQuestionStart(USER_ID, USER_IDX, userContent, answer)) }.start()

                addMsg(answer, UserChatMessage.TYPE_OTHER, false, null, suggests)
                ttsManager.speak(answer)
                Log.d("ServerCheck", "파싱 및 화면 업데이트 성공")
            } else {
                Log.e("ServerCheck", "서버 status가 SUCCESS가 아님: ${root.optString("status")}")
                resetToIdleState()
            }
        } catch (e: Exception) {
            Log.e("ServerCheck", "JSON 파싱 중 에러 발생: ${e.message}")
            e.printStackTrace()
            resetToIdleState()
        }
    }

    private fun startRecordingFlow(manual: Boolean) {
        voskManager.stopListening()
        if (!manual) {
            ttsManager.speak("네, 말씀하세요.")
            mainHandler.postDelayed({
                if (!isRecording && loadingBar.visibility != View.VISIBLE) voiceRecorder.start(5000)
            }, 1200)
        } else {
            voiceRecorder.start(5000)
        }
    }

    private fun openBackCamera() {
        val intent = Intent(MediaStore.ACTION_IMAGE_CAPTURE).apply {
            putExtra("android.intent.extras.CAMERA_FACING", 0)
        }
        try { takePictureLauncher.launch(intent) } catch (e: Exception) { resetToIdleState() }
    }

    private fun resetToIdleState() {
        isRecording = false
        shouldLaunchCamera = false
        updateMicButtonUI(forceMic = true)
        runOnUiThread { if (::voskManager.isInitialized) voskManager.startListening() }
    }

    private fun setBearMood(mood: BearMood) {
        runOnUiThread {
            val res = when (mood) {
                BearMood.BASIC -> R.drawable.ddolddol_basic
                BearMood.WORRY -> R.drawable.ddolddol_worry
                BearMood.PRAISE -> R.drawable.ddolddol_praise
            }
            ivBear.setImageResource(res)
            ivBear.alpha = if (mood == BearMood.PRAISE) 0.25f else 0.13f
        }
    }

    private fun buildUserContext(): String {
        val sb = StringBuilder("[사용자 정보]\n")
        if (userSpecialNote.isNotBlank()) sb.append("특이사항: $userSpecialNote\n")
        if (whitelistTaskNames.isNotEmpty()) {
            sb.append("보호자가 허용한 활동: ${whitelistTaskNames.joinToString(", ")}\n")
        }
        sb.append("사용자는 중등도 지적장애가 있습니다. 짧고 쉬운 말로, 한 번에 한 가지만 말해주세요.")
        return sb.toString()
    }

    private fun handleScheduleIntent(intent: Intent) {
        if (intent.getBooleanExtra("schedule_auto_execute", false)) {
            val scheduleId = intent.getStringExtra("schedule_id") ?: return
            val title = intent.getStringExtra("schedule_title") ?: return
            mainHandler.postDelayed({ startScheduleExecution(scheduleId, title) }, 1500)
        }
    }

    private fun showScheduleDialog(scheduleId: String, title: String) {
        runOnUiThread {
            AlertDialog.Builder(this)
                .setTitle("📅 일정 시간이 됐어요!")
                .setMessage(title)
                .setPositiveButton("하기") { _, _ -> startScheduleExecution(scheduleId, title) }
                .setNegativeButton("나중에", null)
                .show()
        }
    }

    private fun startScheduleExecution(scheduleId: String, title: String) {
        val id = scheduleId.toIntOrNull() ?: -1
        scheduleTitle = title
        currentScheduleId = id
        currentStepIndex = 0
        Thread {
            val (steps, note) = ScheduleRepository.fetchStepsByScheduleId(currentScheduleId, USER_ID, USER_IDX)
            runOnUiThread {
                scheduleSteps = steps
                scheduleNote = note
                sendSchedulePromptToAI()
            }
        }.start()
    }

    private fun sendSchedulePromptToAI() {
        if (loadingBar.visibility == View.VISIBLE || isRecording) return
        voskManager.stopListening()
        val prompt = buildSchedulePrompt()
        addMsg("📅 $scheduleTitle ${stepProgressLabel()}", UserChatMessage.TYPE_MINE, false, null, null)
        uploadToServer(null, null, prompt)
    }

    private fun buildSchedulePrompt(): String {
        val userContext = buildUserContext()
        val noteStr = if (scheduleNote.isNotBlank()) "참고사항: $scheduleNote\n" else ""
        return if (scheduleSteps.isEmpty()) {
            "$userContext\n[일정 수행 모드]\n일정: $scheduleTitle\n${noteStr}일정을 시작합니다."
        } else {
            val stepDesc = scheduleSteps[currentStepIndex]
            "$userContext\n[일정 수행 모드]\n일정: $scheduleTitle\n단계: $stepDesc"
        }
    }

    private fun stepProgressLabel() = if (scheduleSteps.isEmpty()) "시작" else "(${currentStepIndex + 1}/${scheduleSteps.size}단계)"

    private fun proceedToNextStep() {
        if (!isScheduleMode) return
        currentStepIndex++
        if (scheduleSteps.isNotEmpty() && currentStepIndex >= scheduleSteps.size) {
            finishSchedule()
            return
        }
        sendSchedulePromptToAI()
    }

    private fun finishSchedule() {
        ScheduleRepository.completeSchedule(currentScheduleId)
        currentStepIndex = -1; currentScheduleId = -1
        setBearMood(BearMood.PRAISE)
        val doneMsg = "수고했어요! 일정을 완료했어요."
        addMsg(doneMsg, UserChatMessage.TYPE_OTHER, false, null, null)
        ttsManager.speak(doneMsg)
    }

    override fun onStart() {
        super.onStart()
        if (::voskManager.isInitialized) voskManager.startListening()
        val filter = IntentFilter(ScheduleAlarmReceiver.ACTION_SCHEDULE_ALERT)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(scheduleReceiver, filter, Context.RECEIVER_NOT_EXPORTED)
        } else {
            registerReceiver(scheduleReceiver, filter)
        }
    }

    override fun onStop() {
        super.onStop()
        if (::voskManager.isInitialized) voskManager.stopListening()
        try { unregisterReceiver(scheduleReceiver) } catch (e: Exception) { }
    }

    override fun onDestroy() {
        super.onDestroy()
        mainHandler.removeCallbacksAndMessages(null)
        if (::voskManager.isInitialized) voskManager.stopListening()
        if (::voiceRecorder.isInitialized) voiceRecorder.release()
        if (::ttsManager.isInitialized) ttsManager.release()
    }

    private fun updateVoiceStatus(newText: String?) {
        if (voiceMsgIndex != -1 && voiceMsgIndex < chatList.size) {
            chatList[voiceMsgIndex].content = newText
            adapter.notifyItemChanged(voiceMsgIndex)
        }
    }

    private fun addMsg(c: String?, t: Int, i: Boolean, b: Bitmap?, s: MutableList<String?>?) {
        chatList.add(UserChatMessage(c, t, i, b, s))
        adapter.notifyItemInserted(chatList.size - 1)
        rvChat.smoothScrollToPosition(chatList.size - 1)
    }

    private fun checkPermissions() {
        val perms = mutableListOf(Manifest.permission.RECORD_AUDIO, Manifest.permission.CAMERA)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) perms.add(Manifest.permission.POST_NOTIFICATIONS)
        val needed = perms.filter { ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED }
        if (needed.isNotEmpty()) ActivityCompat.requestPermissions(this, needed.toTypedArray(), 100)
        else voskManager.initModel()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 100 && grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) voskManager.initModel()
    }

    override fun onSuggestionClick(text: String?) {
        if (loadingBar.visibility == View.VISIBLE || isRecording || ttsManager.isSpeaking()) return
        if (isScheduleMode && (text == "다음 단계로" || text == "완료")) { proceedToNextStep(); return }
        voskManager.stopListening()
        voiceMsgIndex = -1
        addMsg(text, UserChatMessage.TYPE_MINE, false, null, null)
        uploadToServer(null, null, text)
    }
}